### ISSUE TITLE
Keyword-only arguments are parsed incorrectly

### ISSUE THREAD

**Issue by eric-wieser** [numba/numba#5903]
## Reporting a bug

- [ ] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [ ] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

**Comment by eric-wieser** [comment]
Looks like this needs at least:

* A patch to `_DispatcherBase` to either use `__kwdefaults__` or `py_sig.parameters[name].default`
* A patch to `find_named_args` in the C code to remove the assumption that default arguments must be adjacent.

**Comment by esc** [comment]
@eric-wieser thanks for reporting this, can you see if this is related: https://github.com/numba/numba/pull/5292

**Comment by eric-wieser** [comment]
Mostly unrelated. That's about _allowing_ `foo(*args, kw)`.

I think the core of this issue is that `__defaults__` are handled incorrectly.

Perhaps the thing to do right now is make `Dispatcher.__init__` throw `NotImplementedError` if any arguments are kwargs and any arguments have defaults

**Comment by eric-wieser** [comment]
Ah, however the conversation that follows in that patch _is_ the same as this problem.

**Comment by esc** [comment]
@eric-wieser thanks for looking into this, I would imagine that the two issues might be related.

**Comment by eric-wieser** [comment]
An extensive rewrite could fix both at the same time, probably.

**Comment by esc** [comment]
> An extensive rewrite could fix both at the same time, probably.

Yes! I concur!

**Comment by stuartarchibald** [comment]
I think the immediate task here is to just outright ban situations that produce nonsense, fixes can then be performed in the relative safety of knowing that there's nothing out there half working.

**Comment by eric-wieser** [comment]
Agreed, it's way more important to stop code silently producing wrong results than it is to add support for a wider range of calling conventions - so starting with the extensive rewrite is a bad idea.

**Comment by kmaitreys** [comment]
Any updated on this? My jitted function does not have kwargs followed by a *, but I am trying to use pass my jitted function to ```scipy.integrate.ode``` and it fails with the ```TypeError: not enough arguments: expected 5, got 4``` kind of error, even when I pass all the parameters/argumements through the ```set_f_params``` method

**Comment by ShaiAvr** [comment]
I am experiencing the same issue. I tried to make my jitted function more flexible by allowing the user to specify the output's `dtype` like many `numpy` functions do:
```python
from numba import jit
import numpy as np

jit_opts = dict(
    nopython=True, nogil=True, cache=True, error_model="numpy", fastmath=True
)


@jit(**jit_opts)
def numerical_scheme(max_steps, tolerance, *, dtype=np.float_):
    arr = np.empty(max_steps, dtype=dtype)  # Allow control of dtype
    # Numerical calculations...
    return arr


a1 = numerical_scheme(50, 1e-1, dtpye=np.float32)  # Maybe I don't need as much precision for this calculation
a2 = numerical_scheme(100, 1e-3, dtpye=np.complex128)  # Maybe this calculation involves complex numbers
```
but I got an error `TypeError: not enough arguments: expected 6, got 5`. The only solution is to make `dtype` a regular positional argument. Then I can call `numerical_scheme(..., dtype=np.float64)` successfully, but it gives the option to call `numerical_scheme(..., np.float64)` which I don't like.
